discovery-handler-modules/coap-discovery-handler/src/discovery_handler.rs

```diff
@@ -0,0 +1,257 @@
+        let discover_request = request.get_ref();
+        let (mut discovered_devices_sender, discovered_devices_receiver) =
+            mpsc::channel(DISCOVERED_DEVICES_CHANNEL_CAPACITY);
+        let discovery_handler_config: CoAPDiscoveryDetails =
+            deserialize_discovery_details(&discover_request.discovery_details)
+                .map_err(|e| tonic::Status::new(tonic::Code::InvalidArgument, format!("{}", e)))?;
+
+        info!("Discovering config {:?}", discovery_handler_config);
+
+        let multicast = discovery_handler_config.multicast;
+        let static_addrs = discovery_handler_config.static_ip_addresses;
+        let multicast_addr = discovery_handler_config.multicast_ip_address;
+        let query_filter = discovery_handler_config.query_filter;
+        let timeout =
+            Duration::from_secs(discovery_handler_config.discovery_timeout_seconds as u64);
+
+        tokio::spawn(async move {
+            loop {
+                let mut devices: Vec<Device> = vec![];
+
+                // Discover devices via static IPs
+                for ip_address in &static_addrs {
+                    let coap_client = CoAPClientImpl::new((ip_address.as_str(), 5683));
+                    let device = discover_endpoint(
+                        &coap_client,
+                        &ip_address,
+                        query_filter.as_ref(),
+                        timeout,
+                    );
+
+                    match device {
+                        Ok(device) => devices.push(device),
+                        Err(e) => {
+                            info!("Discovering endpoint {} went wrong: {}", ip_address, e);
+                        }
+                    }
+                }
+
+                // Discover devices via multicast
+                if multicast {
+                    let coap_client = CoAPClientImpl::new((multicast_addr.as_str(), 5683));
+                    let discovered =
+                        discover_multicast(&coap_client, query_filter.as_ref(), timeout);
+
+                    match discovered {
+                        Ok(mut discovered) => {
+                            devices.append(&mut discovered);
+                        }
+                        Err(e) => {
+                            error!("Error while discovering devices via multicast {}", e);
+                        }
+                    }
+                }
+
+                if let Err(e) = discovered_devices_sender
+                    .send(Ok(DiscoverResponse { devices }))
+                    .await
+                {
+                    error!(
+                        "discover - for CoAP failed to send discovery response with error {}",
+                        e
+                    );
+                    register_sender.send(()).await.unwrap();
+                    break;
+                }
+
+                delay_for(Duration::from_secs(DISCOVERY_INTERVAL_SECS)).await;
```

Comment:
I just copied over the implementation from you. Make it work first, then improve it 😂
Jokes aside, 10s works in my experience but it's very personal. Less time can result in the device not being able to reply in time since we are dealing with very constrained devices and unstable networks.

> would lack of a loop and rather a polling for new devices be better?

Would you mind expanding your question? By polling, do you mean sending a request and waiting forever for its response like web clients used to do before WebSockets?

Reply:
I just want to make sure you arent tied to a 10 loop and rather doing whatever form of discovery best fits coap. For example, maybe you maintain the multicast socket and just check if new devices are discovered -- there is the issue of knowing when a device goes offline in this case though.